### ISSUE TITLE
Pluggable storage backends (was Support for Consul K/V storage)

### ISSUE THREAD

**Issue by Galmido** [kubernetes/kubernetes#1957]
[Consul](http://www.consul.io/) is a self checking service discovery and dynamic DNS server which runs on all machines in our bare-metal datacenter. We use Consul to provide fast and reliable DNS resolving service, mainly for Kubernetes Pods addresses. Consul also provides a simple REST K/V storage system which is very similar to Etcd and we like to use it instead whole Etcd cluster.

Currently we use Consul and Etcd together, where Consul K/V is used by all our services, and Etcd is only used by Kubernetes. This forces us to maintain two independent datacenter-wide services which provides almost exactly the same functionality without any additional benefits.

Both services have very similar REST API and it shouldn't be difficult to provide support for Consul.

**Comment by kcq** [comment]
+1.01

**Comment by thomassun** [comment]
@Galmido I'm very interesting in how you combine Kubernetes with Consul to offer pods address because I have just almost the same requirement. But I'm new to Consul, I wonder how Consul know the lifecycle of a pod and  the address(say IP and hostPort) of it. Thanks.

**Comment by bgrant0607** [comment]
FWIW, we've also had a request for Zookeeper.

Things we want out of a k-v store (not all of which are currently delivered by etcd), off the top of my head:
* Optimistic concurrency
* Access to past versions of objects
* Efficient reliable watch
* Efficient recursive or other multi-object watch
* Multi-object transactions
* The ability to somehow build multiple indices
* Pagination support
* Access control
* Low per-object memory footprint
* Object TTLs (for events, logically deleted objects)

**Comment by rkettelerij** [comment]
+1 for consul

**Comment by MrMMorris** [comment]
big +1 for consul

**Comment by bgrant0607** [comment]
This will not be a priority for us in the foreseeable future (see our [roadmap](https://github.com/GoogleCloudPlatform/kubernetes/blob/master/docs/roadmap.md)). However, we'd be happy to work with someone from the community on this, in perhaps a couple months after we simplify how Kubernetes interacts with the storage layer.

**Comment by lavalamp** [comment]
@saad-ali is going to do some etcd benchmarking, which will give us information to determine the urgency of this issue.

**Comment by thockin** [comment]
In terms of overall utility - the value of doing a second storage
implementation is that it will force us to keep our API honest.  I would
love to put a bounty on someone doing a second impl, even if we don't have
time for it on the core team.

**Comment by bgrant0607** [comment]
More specifically what I think we need to finish first:
* Post status back #2726 
* Eliminate BoundPods #2483
* Clean up registry etc. #1451 and make all objects use the generic etcd registry implementation

**Comment by jletroui** [comment]
+1 for consul

**Comment by mainframe** [comment]
+1 for consul

**Comment by offlinehacker** [comment]
+1 consul could be also used as replacement for skydns

**Comment by bartvankleef** [comment]
+1

**Comment by sumanthy** [comment]
+1 consul

**Comment by rombob** [comment]
+1

**Comment by geeknam** [comment]
+1 consul

**Comment by rushmorem** [comment]
:+1:

**Comment by radeksimko** [comment]
:+1:

**Comment by jefferai** [comment]
Big +1 for Consul. We've run into problems with etcd due to bugs in their (now previous) underlying Raft layer. They have written a new Raft layer and yet some of the same bugs are being reported by users, plus it's too new to be fully battle tested.

Kubernetes using etcd is the main reason we're not seriously evaluating Kubernetes. We just don't trust etcd for production systems right now.

On the other hand, we've had excellent experience with Consul. Now that it supports TTLs, Kubernetes being backed by Consul is something we would seriously evaluate.

**Comment by thockin** [comment]
We're certainly in favor of making it work, but I want to point out a
couple things.

First, that this is not something exposed to users.  It is an
implementation detail of our master.

Second, that this is NOT something that the core team is working on right
now.  If you want to see it, please feel free to pitch in :)

**Comment by smarterclayton** [comment]
A good place to start is the generic etcd store implementation - that handles the standard CRUD operations along with watch.  No one has yet done an alternative store implementation, but we're happy to advise and review!

**Comment by jefferai** [comment]
'''
Second, that this is NOT something that the core team is working on right
now.  If you want to see it, please feel free to pitch in :)
'''
Understood, and I might, but my understanding was that there were many things very etcd-specific and that a more abstract layer was somewhere in the pipeline. Take my comment as a vote for a more abstracted interface, if it doesn't already exist.  :-)  That will allow many other backends to be written.

**Comment by a86c6f7964** [comment]
I will vote for "abstracted ability to move to consul in the future" as well. I was pretty unhappy to see comments around "please move to etcd 2.0" because this is probably fixed in that new version when I was having trouble with .4.x of etcd.

**Comment by smarterclayton** [comment]
> On May 1, 2015, at 7:44 PM, Jeff Mitchell <notifications@github.com> wrote:
> 
> '''
> Second, that this is NOT something that the core team is working on right
> now. If you want to see it, please feel free to pitch in :)
> '''
> Understood, and I might, but my understanding was that there were many things very etcd-specific and that a more abstract layer was somewhere in the pipeline. Take my comment as a vote for a more abstracted interface, if it doesn't already exist. :-) That will allow many other backends to be written.
> 
That abstract layer exists today, but there's probably some loose ends dangling around.

**Comment by bgrant0607** [comment]
Much of the previously mentioned refactoring has been done. The genericetcd registry mainly needs to be made more abstract. Take a look:
https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/registry/generic/etcd/etcd.go

**Comment by thockin** [comment]
The only way to really have an abstraction is to have more than one
concrete implementation. [Your project name here] could be the first
alternate implementation!

**Comment by alex-mohr** [comment]
From reading the Jan 13 comment by @bgrant0607 it seems like what we're looking for smells more like a database than a plain k-v store.  If we're going to look at alternate storage implementations, can we please do a due diligence on "real databases"?

**Comment by CpuID** [comment]
+1 for Consul

**Comment by bhourigan** [comment]
+1 for consul

**Comment by omribahumi** [comment]
+1 for consul

**Comment by jasonkeene** [comment]
+1 for consul

**Comment by umrigark** [comment]
+1 for Consul

**Comment by rcousens** [comment]
+1 for consul

**Comment by ncode** [comment]
+1 for consul

**Comment by rogerlz** [comment]
+1 for consul

**Comment by a86c6f7964** [comment]
I didn't want to do this because it just adds to the problem. But I am
watching this issue to be able to see if there is progress on this issue
and all of the +1's going to everyone's email who watches this issue, I
don't think is going to help.

**Comment by wienczny** [comment]
Did you have a look at https://github.com/docker/libkv
It its an abstraction layer for Etcd, Consul, Zookeeper
While it does not support all features of consul (e.g. services) it could be useful.

**Comment by naelyn** [comment]
I started pushing bits around to play with what this would entail.  If we ignore the terminology aspect of the refactor ("etcd" shows up in a lot of places and variable names where "kvstore" would fit just fine), there seems to be just a bit of an impedance mismatch about TTLs between etcd and consul and how k8s uses TTLs.

Etcd TTLs are like Memcached/Redis TTLs and are "owned" by the key (the TTL labels the key).  The remote server tracks all of the keys in something like a giant priority queue and kicks them out when they get too old.

Consul TTLs are self-described to be like Chubby Sessions.  You create a new entity called a Session, you assign it a TTL and a behavior for what happens when the TTL runs out (for k8s you'd want to use "delete" behavior).  Then you stamp that session (via a lock acquisition action) onto the keys in the KV store if you want them to share a fate with the Session.  The consul leader keeps all of the sessions in a priority queue and evicts the ones that are too old. This cuts down on the memory footprint on the elected Consul leader as in the desired case (compared to etcd), there are fewer sessions to track than keys.

Unfortunately, if you try to marry the etcd-flavor and consul-flavor of kv-with-ttls things start to get weird.  You currently would have to:

1. create a session with the ttl and behavior=delete (1 rpc, 1 raft op)
2. persist the KV pair with the session lock acquired (1 rpc, 1 raft op)
3. reread the data that was saved (since the k8s code seems to rely upon etcd writes returning the written data) (1 rpc)

Also AFAICT you cannot both do a CAS operation and a lock acquisition atomically together.  You have to sequence them.  I can't tell how often in the k8s codebase it relies upon CAS & TTL.

It is not impossible to contort Consul primitives to make TTL'ed edits to keys work like k8s expects, but it gets a little bit questionable from the perspective of proving that the final result is actually *safe*.

**Comment by erictune** [comment]
I did some grepping and it looks like etcd TTLs are currently only used to garbage collect `event` objects.
See use of `TTLFunc` in `pkg/registry/event/registry.go`

I suspect we would be open to a patch to remove the use of the etcd TTL functionality, replacing it with a background GC goroutine.

**Comment by smarterclayton** [comment]
We've made the statement that we would not expand the use of TTL for
graceful deletion (because there are other downsides to using it) so it's
safe to say we would not extend the use of TTL at this point.

On Wed, Aug 5, 2015 at 2:05 AM, Eric Tune <notifications@github.com> wrote:

> I did some grepping and it looks like etcd TTLs are currently only used to
> garbage collect event objects.
> See use of TTLFunc in pkg/registry/event/registry.go
>
> I suspect we would be open to a patch to remove the use of the etcd TTL
> functionality, replacing it with a background GC goroutine.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/issues/1957#issuecomment-127882814>
> .
>



-- 
Clayton Coleman | Lead Engineer, OpenShift

**Comment by Niyakiy** [comment]
+1
It would be nice to completely move from etcd to consul.

**Comment by naelyn** [comment]
Following the comment by @erictune, I mostly have a POC of the background garbage collector written. It's not in a completed state yet, but the branch is here: https://github.com/naelyn/kubernetes/commits/no-ttl-events

Probably the order of operations is:

1. switch from TTL to background GC
2. figure out if anything else is still using TTL and deal with it
3. create dual support abstractions for storage (etcd, consul) making sure to hide all of the etcd-isms (like specific errors used programmatically downstream)
4. feature parity in the testing space

I had hoped that the docker/libkv library would help provide a nice rosetta stone of quirk-parity, but unfortunately I already found some incongruities in that library so it's probably best to start fresh on that front.

**Comment by bgrant0607** [comment]
There's also discussion of a garbage collector (for pods) on #7660. cc @nikhiljindal

**Comment by erictune** [comment]
@naelyn 
It has to be possible to upgrade cluster to a version that has background GC and then downgrade back to one that doesn't have it.  

So, we might need to do both for 1 minor version, and then turn off TTL in the next minor version.

**Comment by anuaimi** [comment]
+1

**Comment by aronchick** [comment]
I'd love to ask all the folks who have +1'd what the scenario you're looking for is. Is it that you're looking for the services as an API endpoint that @bgrant0607 talked about? E.g. a k-v store that does:

* Optimistic concurrency
* Access to past versions of objects
* Efficient reliable watch
* Efficient recursive or other multi-object watch
* Multi-object transactions
* The ability to somehow build multiple indices
* Pagination support
* Access control
* Low per-object memory footprint
* Object TTLs (for events, logically deleted objects)

Or is it something else? That is to say, to echo @thockin's point, etcd is internal to Kubernetes, and is not made available as an end-user service. 

Additionally, Kubernetes does provide service discovery in the form of labels, and health-checking in each replication controller - is there more that you're looking for from Consul?

**Comment by CpuID** [comment]
@aronchick one of the motivators is not wanting to maintain an etcd cluster for production use, when we already have a consul cluster for other purposes.

**Comment by emaildanwilson** [comment]
+1 My thought on this is if it integrates into an external service registry then accessing external resources can be performed without adding services. Also, and possibly more important, when services running inside a single k8s cluster try to access another service it could auto-route to another copy of the service in another k8s cluster if the local one is down (still thinking same region here). Lastly, and least important, I think w/ consul as a replacement for etcd you could further simplify k8s by removing skydns, correct?

**Comment by aronchick** [comment]
Thanks @CpuID & @emaildanwilson - I can't comment specifically about skydns, but, in this case, etcd does not have to be maintained by you, this is an internal service that k8s uses.

We are definitely thinking about cross-cluster service resolution, though that's part of the federation proposal - https://github.com/kubernetes/kubernetes/blob/master/docs/proposals/federation.md cc @quinton-hoole

**Comment by jefferai** [comment]
@aronchick Cross-DC resolution is going to be in Consul soon. You'd save yourselves a lot of trouble!  :-)

**Comment by aronchick** [comment]
@jefferai we'd love to understand more - how do you see Consul working in the context of Kubernetes labels?

**Comment by bgrant0607** [comment]
Using consul for DNS is not what this issue is about. If you'd like to publish endpoints to consul, that should get its own issue, but I don't think anything prevents anyone from writing an addon that does exactly that. Feel free to send a PR to the contrib repo.

**Comment by aronchick** [comment]
cc @jefferai

**Comment by nmabhinandan** [comment]
We use consul for our (micro) services because it's better than etcd. And we don't wanna run separate etcd cluster just for kubernetes. So +1 for consul.

**Comment by jefferai** [comment]
@aronchick I don't really understand the question; Kubernetes labels are K/V entries so that seems to just rely on having an underlying K/V capability.

Maybe what you wanted to understand more about is Consul and its service discovery capabilities, and I misunderstood the question.

**Comment by aronchick** [comment]
@jefferai Hey Jeff - thanks for writing back. Consul solves service discovery, uses an agent, etc. - all very powerfully. 

But Kubernetes solves this a different way - it uses a label as a selector, and allows a user to expose that as a service endpoint with no agent. The fact that it uses a K/V store (whatever it might be) under the hood is totally hidden from the user.

For example, let's say you started a service endpoint per the example on this page (https://consul.io/intro/getting-started/services.html). You would install an agent in the container, create a service document in the container (/etc/consul.d) and run the agent to register the service :

``` 
$ consul agent -server -bootstrap-expect 1 -data-dir /tmp/consul \
    -config-dir /etc/consul.d
==> Starting Consul agent...
...
    [INFO] agent: Synced service 'web'
...
```

This would result in the following:

```
dig @127.0.0.1 -p 8600 web.service.consul

;; QUESTION SECTION:
;web.service.consul.        IN  A

;; ANSWER SECTION:
web.service.consul. 0   IN  A   172.20.20.11
```

Kubernetes works differently - instead of using an agent to register at run time, the pod is automatically registered as a service endpoint just by having a label. That is to say, if I start a pod in the following way:

```
cluster/kubectl.sh run web-rc --image=my-sweet-web-image:latest --label=web --replicas=20
```

And created a service endpoint:

```
cluster/kubectl expose rc web-rc --port=80 --selector="web" --name=web-endpoint
```

The exact same query would work in the container:

```
$ # Use Exec to get into one of the running pods
$ cluster/kubectl.sh exec web-3ljt5 -ti bash
root@web-3ljt5:/app# dig +search web-endpoint

; <<>> DiG 9.9.5-3ubuntu0.2-Ubuntu <<>> +search web-endpoint
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 63208
;; flags: qr aa rd ra; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 0

;; QUESTION SECTION:
;web-endpoint.default.svc.cluster.local. IN A

;; ANSWER SECTION:
web-endpoint.default.svc.cluster.local. 30 IN	A 10.0.25.158

;; Query time: 2 msec
;; SERVER: 10.0.0.10#53(10.0.0.10)
;; WHEN: Tue Sep 01 00:45:25 UTC 2015
;; MSG SIZE  rcvd: 74
root@web-3ljt5:/app#
```

You wouldn't use any of the higher function capabilities of Consul - all those (load balancing, health checking, etc etc) are all provided by Kubernetes.

Again, the point is that you don't have to ever touch etcd - it is only an internal service.

**Comment by nmabhinandan** [comment]
@aronchick Yes. But we have to install etcd just for Kubernetes. It would be great if there was an abstraction layer which supports multiple kv store backends.
We're already stuck with 3 systems.
Zookeeper for mesos, kafka, etc. 
consul for our app microservices
etcd for kubernetes, weave, calico etc.
It's a pain in the ass to maintain several clusters for same kinda purpose.

**Comment by CpuID** [comment]
> It's a pain in the ass to maintain several clusters for same kinda purpose.

Exactly what @nmabhinandan said

**Comment by davidopp** [comment]
Whatever key/value store we use needs to support watch to work today. And in the future, we plan to take advantage of multi-object transactions (which etcd will support early next year according to their roadmap). We're getting rid of our use of TTLs so that's not needed.

ZK appears to support watch and multi-object transactions. It appears to have a Go client library created by a community member.

Consul appears to support watch but doesn't seem to support multi-object transactions. It's written in Go and appears to have an official Go client library.

Two other features @bgrant0607 mentioned earlier were optimistic concurrency control and access to past versions of objects (e.g. ability to watch starting in the past). Both appear to support optimistic conccurency control. I'm not sure if they support access to past versions of objects.

**Comment by aronchick** [comment]
@nmabhinandan @CpuID Can you say more about how this fits into the other KV stores? To your point, you _wouldn't_ maintain etcd at all for Kubernetes. Just think of it like libc - it's there, and it does stuff, but it's not part of anything you'd have to worry about. You would never have to watch it, monitor it, restart it, upgrade it, etc etc.

**Comment by lavalamp** [comment]
To those of you who want to reuse an existing consul installation for kubernetes: we have been generally recommending that no one share kubernetes' data store. Two reasons are: Kubernetes puts a lot of load on its backend, and security: e.g. the secrets that kubernetes stores are sensitive. Access to the storage backend == root on your cluster.

**Comment by xiang90** [comment]
@jefferai 

> Big +1 for Consul. We've run into problems with etcd due to bugs in their (now previous) underlying Raft layer. They have written a new Raft layer and yet some of the same bugs are being reported by users, plus it's too new to be fully battle tested.

As an etcd maintainer, I can say this situation has changed. We have done a lot of tests for etcd (More precisely we injected 1000s of failures into etcd clusters we are running everyday.) If you meet any issues, please let us know. We received almost zero bug reports related our new raft from our users since the very beginning actually.

Also as a k8s contributor, I do not really think we need the functionality (service discovery, Cross-DC resolution) you mentioned above at the moment. I think @aronchick @bgrant0607 has well stated what k8s needs. From my understanding none of the existing open source solution is a prefect fit right now. But any data store goes towards to that direction would be a huge help for k8s community.

Thanks!

**Comment by jedsmith** [comment]
@aronchick

> You would never have to watch it, monitor it, restart it, upgrade it, etc etc.

What do you mean by this? You absolutely have to do all of those things with etcd to run *and* scale Kubernetes.

**Comment by bgrant0607** [comment]
Also note that we're considering moving high-traffic writes of soft state, such as to status, events, and metrics, to another, in-memory backend -- see #6363.

**Comment by bgrant0607** [comment]
I also don't agree with @aronchick that the storage backend wouldn't have to be monitored, upgraded, etc. I do, however, agree with @lavalamp that we recommend not sharing the storage backend with other clients.

**Comment by nmabhinandan** [comment]
> We recommend not sharing the storage backend with other clients. - @bgrant0607 

Slightly off topic but shouldn't I use the same etcd cluster for kubernetes and my for own applications? Why?

**Comment by davidopp** [comment]
> Slightly off topic but shouldn't I use the same etcd cluster for kubernetes and my for own applications? Why?

https://github.com/kubernetes/kubernetes/issues/1957#issuecomment-137270726

**Comment by jefferai** [comment]
@aronchick I think there is some confusion going on in our discussion, so let me try to clarify:

> But Kubernetes solves this a different way - it uses a label as a selector, and allows a user to expose that as a service endpoint with no agent. The fact that it uses a K/V store (whatever it might be) under the hood is totally hidden from the user.

My point was that because the K/V store is under the hood, whether that is powered by Consul or etcd or ZK shouldn't matter so long as the storage solution supports the K/V store requirements for Kubernetes labels.

> For example, let's say you started a service endpoint per the example on this page (https://consul.io/intro/getting-started/services.html). You would install an agent in the container, create a service document in the container (/etc/consul.d) and run the agent to register the service...Kubernetes works differently - instead of using an agent to register at run time, the pod is automatically registered as a service endpoint just by having a label. 

If you're using the service discovery capabilities of Consul -- which again, are not a requirement for the K/V store -- Consul *can* run with an agent on each node, but it is *not* required to. You can simply use the API (example at https://gist.github.com/jefferai/7461cce52179b22dae5e). The agent is for anti-entropy -- it allows the system to automatically prune services out of the catalog if e.g. the agent is down, or stops reporting a particular service that it originally injected into the system.

> You wouldn't use any of the higher function capabilities of Consul - all those (load balancing, health checking, etc etc) are all provided by Kubernetes.

> Again, the point is that you don't have to ever touch etcd - it is only an internal service.

Right, which goes right back to my point. You don't need any of the higher function capabilities of Consul to use it as a K/V store. If all Kubernetes labels needs is a K/V store, then it should fit the bill just fine. And, after all, this ticket is about using Consul as a Kubernetes K/V store  :-)

There are other issues on the tracker related to bringing in SkyDNS and a reliance on etcd there where it'd be very nice to be able to hook into working, highly available Consul setups if available rather than have a secondary DNS/service discovery setup. But that's not really what this ticket is about.

**Comment by jefferai** [comment]
@lavalamp Of course, that's what https://github.com/kubernetes/kubernetes/issues/10439 is about, which has turned partly into a discussion of storing secrets in Vault, and partly a discussion of simply encrypting secrets within Kubernetes so that it doesn't matter what storage backend is used. But then Kubernetes has to manage the master secrets, and that's a hard problem where Vault can be helpful too.

**Comment by thockin** [comment]
@jefferai 

> after all, this ticket is about using Consul as a Kubernetes K/V store :-)

The part I don't understand is why people care?  If the K/V store is a) not visible to users, b) not providing direct service to users, and c) not using higher-order features of the underlying implementation - why does anyone care whether it is etcd, consul, or something else?

My understanding of this bug was mostly that people care because the ALREADY have consul and want to use and share that instance, which we really discourage, and that people want higher level tie-ins.  Using higher-level features is easy enough - just write a bridge that syncs k8s API into consul, the same way we do for our DNS server (which has a tiny etcd instance).

**Comment by a86c6f7964** [comment]
My concern is just that I have been able to have a stable consul cluster running in prod for a long time, where as I had to ditch the etcd1 cluster because it was causing my machines to have problems. I have not tried etcd2 yet. That being said, does k8s manage running the etcd cluster itself? I would not care if that were the case, but I think that I would assume that you configure the etcd cluster that k8s talks to? Which just means that its another operational thing to understand, if I already understand how to run consul.

**Comment by thockin** [comment]
We own the etcd cluster.  That it exists is an implementation detail :)

**Comment by jefferai** [comment]
@thockin People care because of PCI, SOX, HIPPA, SLAs, and all sorts of other reasons that require strong controls and documented practices around where data is being stored, how it is being backed up, its supposed level of service, and so on. It's not dissimilar to reasons that enterprisey types don't run internal sqlite databases for applications. They connect them to a managed, replicated, monitored, clustered, backed up MySQL or Postgres or Maria or Oracle or some other database that checks all the boxes required for legal or compliance or even internal policy reasons.

You guys discourage using a shared instance because of write load (which may not actually be a problem if an instance is not otherwise heavily loaded, plus other KV stores are faster than etcd) and because of secrets being stored in the KV store. What you're missing is that that's *exactly why* people want to connect Kubernetes to a shared instance...so that load and monitoring and access to those secrets can be managed according to policies in effect for that organization, not simply policies decided by the Kubernetes team. And if you've already gone through the entire vetting and management and monitoring setup and all of those things with Consul, and Consul supports the necessary functionality to act as Kubernetes' KV store, then it's not surprising that people would prefer to connect Kubernetes to it than have Kubernetes use something internally.

Access to higher order functions is really not the issue here. But the people working on higher-order functions within Kubernetes are having similar problems understanding the issues. A tiny, single-node etcd instance is a single point of failure for service discovery, which is simply an unacceptable risk in many enterprises. Whether you solve that by having SkyDNS on a multi-node etcd instance, or by abstracting it and letting DNS be handled by Consul, is really kind of missing the point.

Another example is the current discussion around using Vault as a secret store. The trend seems to be to encrypt the values going into the KV store -- which I think is perfectly sane, and will work regardless of KV store being used -- but then you have to manage the master keys, and that is something that is rife with differing requirements. That's why people are looking into Vault, is that it has put serious effort into solving those use-cases (I've put in a note on the relevant ticket about Vault storing the master keys possibly being a good scenario).

The common thread throughout all of this -- and this is all just as it appears to me, so I may well be wrong -- is that it seems that Kubernetes repeatedly shies away from abstraction, thinking that opinionated functionality is far superior to flexibility. I've actually seen comments from Kubernetes people to this effect but I really can't find the right GitHub issue right now to link to...but the sentiment was basically (paraphrasing) "any time that you start thinking about keeping things abstract you just get everyone wanting their pet solution...just look at that Consul KV store ticket, it was a bad idea to entertain that possibility because now people think other things should be swappable".

From a project maintenance perspective this is undoubtedly true. But it comes down to whether you are trying to build the solution that works the way you currently want, and only the way you currently want, or the solution that -- while still properly designed and architected -- works for a variety of potential users/customers. Right now it doesn't seem like there's much interest in the latter. But while that removes a time/maintenance/support cost, it comes with costs of its own.

**Comment by jonmoter** [comment]
+1 to what @jefferai said, especially around the first 3 paragraphs on controls, SLAs, compliance, etc.

I don't have a strong opinion on the "common thread" on abstraction vs. opinionated functionality. I like opinionated frameworks. But I also recognize that for every additional technology introduced into an ecosystem, there's work required to handle replication, security, compliance, recovery, etc.

**Comment by Hokutosei** [comment]
+1 to Consul. the fact that it has a UI manager, I can do more for the KV.

**Comment by thockin** [comment]
@jefferai ok, there's some fair points in there.  I think the charge that we don't want things to be swappable is patently false, though - we have jumped through hoops to make things swappable and pluggable.  This is an area where we've laid the groundwork but have neither the time nor expertise to make  Consul work (and keep it working) but we'd welcome contributions.

**Comment by bgrant0607** [comment]
Yes, this issue has a help-wanted label on it for a reason: We'd be happy for someone to take this up. But we're swamped and won't get to it any time soon. Just to contrast: People who wanted support for their storage, networking, identity, or auth solutions stepped up to make it happen.

**Comment by jefferai** [comment]
@thockin I do recognize that work that has gone into abstraction to lay the groundwork for other KV stores...I've looked at the code and I'm not sure it's completely ready for other stores (when I looked a few months ago I still saw etcd-specific stuff), but if not it's certainly close. But other decisions that have been made more recently (such as DNS/service discovery) have not (AFAICT) followed that trend, and when you see comments by Kubernetes people lamenting the fact that the work was done to make things more swappable on the KV front because now everyone wants their custom whatever, it doesn't give you a warm fuzzy feeling.

@bgrant0607 My point wasn't that you guys aren't implementing it, or implementing it fast enough. I was trying to help you guys understand why people care, since that is explicitly (https://github.com/kubernetes/kubernetes/issues/1957#issuecomment-137578732) something that was not being understood. If you understand why customers or potential customers have specific desires, you can think about those needs not just for this particular ticket, but also going forward when other technology decisions are being made. (Also, I don't think it's quite credible to claim that people that wanted other storage solutions stepped up to make it happen; browsing around the source code and documentation I only see etcd.)

**Comment by smarterclayton** [comment]
I'll speak to service discovery directly - Kube offers (or attempts to
offer) a holistic view of how service discovery and DNS and interconnection
should perform.  So it's not that the Consul service discovery mechanism is
incompatible, or that we don't see value, it's that we are trying to build
something that is a fundamental building block there, and it's hard to see
how you can have two fundamental building blocks in many cases.  We don't
want to prevent people gluing those together at all.  But we are trying to
make SD consistent in itself if you get Kube without consul.

But!  Someone should definitely spend some time looking to integrate the
two models, but it just so happens that our existing contributors aren't
personally invested in Consul so they don't necessarily know what to
suggest.  Having folks from the Consul community more involved in those
discussions (beyond this issue of KV storage) would be very fruitful, but
we just aren't experts and are likely to miss the insights those folks can
bring.

There are definitely advantages consul SD has today, specifically the way
gossip can spread the health information of services.  A very concrete
direction of exploration...

On Sep 4, 2015, at 8:26 AM, Jeff Mitchell <notifications@github.com> wrote:

@thockin <https://github.com/thockin> I do recognize that work that has
gone into abstraction to lay the groundwork for other KV stores...I've
looked at the code and I'm not sure it's completely ready for other stores
(when I looked a few months ago I still saw etcd-specific stuff), but if
not it's certainly close. But other decisions that have been made more
recently (such as DNS/service discovery) have not (AFAICT) followed that
trend, and when you see comments by Kubernetes people lamenting the fact
that the work was done to make things more swappable on the KV front
because now everyone wants their custom whatever, it doesn't give you a
warm fuzzy feeling.

@bgrant0607 <https://github.com/bgrant0607> My point wasn't that you guys
aren't implementing it, or implementing it fast enough. I was trying to
help you guys understand why people care, since that is explicitly (#1957
(comment)
<https://github.com/kubernetes/kubernetes/issues/1957#issuecomment-137578732>)
something that was not being understood. If you understand why customers or
potential customers have specific desires, you can think about those needs
not just for this particular ticket, but also going forward when other
technology decisions are being made. (Also, I don't think it's quite
credible to claim that people that wanted other storage solutions stepped
up to make it happen; browsing around the source code and documentation I
only see etcd.)

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/issues/1957#issuecomment-137722207>
.

**Comment by bgrant0607** [comment]
@jefferai I appreciate your patience in explaining your desire for the feature.

By "storage", I meant network-attached storage solutions, like NFS. Sorry for the confusion.

**Comment by bgrant0607** [comment]
It's certainly reasonable to discuss what Consul name resolution / service discovery has to offer, but we should split that into a different issue.

**Comment by thockin** [comment]
@jefferai Thanks for a thoughtful discussion.

> I'm not sure it's completely ready for other stores 

I promise it's not - until you have multiple implementations of an abstraction there are ALWAYS leaks.

> other decisions that have been made more recently (such as DNS/service discovery) have not (AFAICT) followed that trend

Do you mean the discussion about internalizing DNS?  The problem with having DNS be an add-on is that many pieces of the system actually expect it to work, and yet probably half of the install instructions out there (including some in-tree!) forget to set up DNS.  It has been a constant source of trouble reports and user hand-wringing - even as recently as THIS MORNING.  Moving it into the core of kubernetes means that we can make the confusion go away, but it means that you don't get to choose which DNS server you run.  We could MAYBE make that pluggable, but the RoI seems pretty low at this point.  DNS is a pretty well-defined interface and we have a reasonably strictly defined schema for kubernetes.  Of course, nothing stops you from running another DNS server and using that instead, but it would not be an out-of-the-box option - it's too niche.

If someone offered a better implementation of DNS - faster, more featureful, more secure, whatever, we would 100% consider it.

> when you see comments by Kubernetes people lamenting the fact that the work was done to make things more swappable on the KV front because now everyone wants their custom whatever, it doesn't give you a warm fuzzy feeling.

I don't know where that reference comes from, but I'll apologize for it anyway.  We (the core team and leadership) are 110% in favor of pluggable stuff.  That said, we try to spend time where it matter the most, and replacing our (currently totally invisible) KV store implementation seems to be low impact.  That said, you've certainly given us some food for thought.  But as before - this will happen if/when someone steps up to help - we're just SWAMPED with a zillion ideas, bugs, features, docs, test. tools, examples, etc to write.

> I don't think it's quite credible to claim that people that wanted other storage solutions stepped up to make it happen

I think Brian meant volumes - we (core team) wrote 3 or 4 and now we have north of a dozen because people stepped in to write their own.  That said, a volume driver is a LOT less work than an adapter for a KV store and we did the hard work of going from 1 impl to 2, so people could easily move from 2 to N.

**Comment by lavalamp** [comment]
To add to what Tim said-- we have a ton of stuff to do, and making the storage pluggable hasn't risen to the top yet. But as we evolve the codebase, we try to evolve it in ways that will make that easier, not harder. See for example the recent caching stuff, where @wojtek-t refactored to make a `storage.Interface`. I'm sure that's not sufficient to get a second system running, but it lowered the surface area a lot.

**Comment by emaildanwilson** [comment]
For my particular use case a plugin to advertise services outside of kubernetes to Consul would work. Do folks think it would be better to write this as a specific plugin for consul or is it better to leverage something like [registrator](https://github.com/gliderlabs/registrator) to provide this capability for multiple service registry type systems as an abstract? Thank you @ghoffer for finding this.  It appears to be written for docker specifically so raising it up a level would require a bit of work. Maybe there is a way to only leverage the libraries and still tying into the project for updates.

**Comment by thockin** [comment]
Having a kube2consul (a la kube2sky) would be the right pattern, I think.

**Comment by jefferai** [comment]
@thockin @bgrant0607 @lavalamp Thanks all for your replies and the discussion. I think we're all on the same page -- I'm not expecting anyone from Kubernetes to go and write a storage driver and I understand why it's low priority and definitely understand having a zillion feature requests to deal with. I just hope that you'll keep some of the food for thought around in terms of enterprise-y type needs.

Regarding service discovery, Kubernetes offering things in a consistent package from core is definitely understandabl. Setting up DNS on your own can certainly be problematic; however, since Consul does it automatically (like SkyDNS) it can probably be performed by a `kube2consul` as @thockin said. I just wanted to sort of make that same point about different (especially enterprise-y) needs and to please keep that in mind when designing interfaces, so that a `kube2consul` drop-in for `kube2sky` is even possible!  :-D

**Comment by aronchick** [comment]
@jefferai To that end, you have the passion and knowledge, and work for Hashi, any chance we could get one of you or your colleagues to start the PR/review? Pretty please? We'll give you all the Google queries you want for free.

**Comment by jefferai** [comment]
@aronchick I honestly can't tell you whether we'd do it or not, but certainly people are aware of this ticket.

Also, just to clarify, I only started at HashiCorp < 2 months ago, so I didn't work for HC when I was first commenting on this ticket, and the enterprise perspective I was talking about earlier in the thread is coming from my previous job. Nothing I've said on this ticket was with my HC hat on. Just want to head off any "of course he's saying that, he's working for HC" thoughts people might be having.  :-)

**Comment by aronchick** [comment]
@jefferai for sure, wasn't implying that. Just that you clearly know the space, and we'd love the help.

**Comment by timothysc** [comment]
just giving this a poke as @wojtek-t and myself have been slowly "exorcising" the code base to enable multiple backends which should be predicated on leveraging the storage abstraction layer.

**Comment by davidopp** [comment]
@feihujiang

**Comment by giorgenes** [comment]
+1 for consul

**Comment by leslie-wang** [comment]
+1 for consul. Asa platform, Kubernetes should not bind to one particular KV store.

**Comment by timothysc** [comment]
Storage is now isolated, for folks who would like to contribute a Consul back-end I can help walk them through what needs to be done and where.  You can hit me up on slack. 

/cc @armon

**Comment by jmelfi** [comment]
@timothysc This is great news!  I'll try to get the details later this week on what needs to happen.

**Comment by shenshouer** [comment]
+1 for consul

**Comment by iT2afL0rd** [comment]
Big +1 for consul

**Comment by deimosfr** [comment]
+1

**Comment by timothysc** [comment]
For those looking to do this work start with creating an equiv backing behind the storage [interface](https://github.com/kubernetes/kubernetes/blob/master/pkg/storage/interfaces.go).  The existing etcd work + the new etcd3 work serves as a good example. https://github.com/kubernetes/kubernetes/tree/master/pkg/storage

**Comment by kipotz1986** [comment]
+1 for consul.

**Comment by trong** [comment]
I do not know about Zookeeper, but Consul is really must have

**Comment by thockin** [comment]
PLEASE do not +1 this thread - use the thumbs-up icon on the first message instead.  Every +1 generates email.

**Comment by steakknife** [comment]
ZK is worth considering for [ginormous](http://www.apache.org/foundation/thanks.html) enterprise / JVM / Hadoop shops.

**Comment by DanyC97** [comment]
@thockin who is the owner of this PR ? would be possible to increase the priority from P3 ? is it reasonable to expect it in say 1.4 or maybe 1.5 time frame ?

Much thanks !

**Comment by thockin** [comment]
This is an issue, not a PR.  It is P3 because it is not critical to any launch plans, and because none of the core maintainers (AFAIK) are working on it.  Boosting the priority is not going to speed it up - we need people to implement it.

**Comment by blalor** [comment]
Implemented by #31622?

**Comment by DocMerlin** [comment]
+1 consul

**Comment by davidopp** [comment]
CoreOS benchmarked etcd, Consul, and ZK. (I was not involved, just saw an announcement and thought folks here might be interested.)

https://coreos.com/blog/performance-of-etcd.html

**Comment by jedsmith** [comment]
@davidopp That benchmark somewhat hints that etcd is evolving into Kubernetes-specific software, which I suppose is unsurprising.

**Comment by davidopp** [comment]
Seriously I wasn't trying to troll. :) I think it would be great if Kubernetes could support multiple key-value stores.

**Comment by smarterclayton** [comment]
I actually don't think etcd is evolving into Kubernetes specific software - we just have continually refined it at fairly large scale, and the etcd team has been phenomenal at tuning etcd to eliminate scalability gaps.

I do think the benchmark numbers mirror our own experiences in production, which it is an excellent high performance and reliable K/V store for O(100k-5m) small keys that also provides the right APIs to run Kube's deliberately simple resource structures.

**Comment by bgrant0607** [comment]
I commented here:
https://github.com/kubernetes/kubernetes/pull/31622#issuecomment-308338309

**Comment by bgrant0607** [comment]
Updated viewpoint:
https://github.com/kubernetes/kubernetes/pull/31622#issuecomment-327606218

See also #46351.

We are not ready to do this in the near future, so I'm actually going to close this for now.

**Comment by binq** [comment]
wow

**Comment by cornfeedhobo** [comment]
Just wow. I've been checking in on this issue, and it's related issues and PRs, for 6 years.

I guess I'll see you all in 2024 at my next check in! :joy: 

![xkcd - time cost](https://imgs.xkcd.com/comics/efficiency.png)

**Comment by o6uoq** [comment]
@cornfeedhobo welcome to open source!

**Comment by yuriy-yarosh** [comment]
It takes forever... I think it would be easier to develop a etcd-consul bridge instead of getting through all the CNCF bureaucracy.

**Comment by lavalamp** [comment]
The last few comments here seem to have misunderstood: it's not the case that we can't decide what to do, it is the case that we have decided not to expand to other storage backends, and every time it comes up we've determined that that's still the right thing for the project, and I don't really see that changing.

Fortunately you don't have to wait for us to change our minds, e.g. check out: https://github.com/k3s-io/kine
